From 542b1c86d9ab0dffc4815cad9086b6164dc0b4ac Mon Sep 17 00:00:00 2001
From: Alexander Duyck <alexander.h.duyck@redhat.com>
Date: Fri, 31 Oct 2014 19:06:27 -0400
Subject: [net] gre: enable offloads for GRE

Message-id: <20141031190627.8251.44934.stgit@ahduyck-vm-rhel7>
Patchwork-id: 98655
O-Subject: [RHEL7.1 net PATCH 3/7] GRE: enable offloads for GRE
Bugzilla: 1151886 1152368
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>
RH-Acked-by: Hannes Frederic Sowa <hannes@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Bugzilla: 1151886
Brew Build: 8179440

Backport of upstream commit: d0a7ebbc119738439ff00f7fadbd343ae20ea5e8

 commit d0a7ebbc119738439ff00f7fadbd343ae20ea5e8
 Author: Amritha Nambiar <amritha.nambiar@intel.com>
 Date:   Thu Jul 10 17:29:21 2014 -0700

    GRE: enable offloads for GRE

    To get offloads to work with Generic Routing Encapsulation (GRE), the
    outer transport header has to be reset after skb_push is done. This
    patch has the support for this fix and hence GRE offloading.

    Signed-off-by: Amritha Nambiar <amritha.nambiar@intel.com>
    Signed-off-by: Joseph Gasparakis <joseph.gasparakis@intel.com>
    Tested-By: Jim Young <jamesx.m.young@intel.com>
    Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Alexander Duyck <alexander.h.duyck@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/gre_demux.c b/net/ipv4/gre_demux.c
index 44e7918..9c8333f 100644
--- a/net/ipv4/gre_demux.c
+++ b/net/ipv4/gre_demux.c
@@ -68,6 +68,7 @@ void gre_build_header(struct sk_buff *skb, const struct tnl_ptk_info *tpi,
 
  skb_push(skb, hdr_len);
 
+ skb_reset_transport_header(skb);
  greh = (struct gre_base_hdr *)skb->data;
  greh->flags = tnl_flags_to_gre_flags(tpi->flags);
  greh->protocol = tpi->proto;
-- 
1.7.1